Output column number info when panicking

est31 opened the pull request:
Outputs the column number when panicking. Useful when you e.g. have code like `foo[i] = bar[k] + bar[l]` and you get a panic with index out of bounds, or when you have an expression like `a = b + c + d + e` and the addition overflows. Now you know which operation to blame!

The format is `file:line:column`, just like for compiler errors. Example output with the patch:

```
thread 'main' panicked at 'index out of bounds: the len is 5 but the index is 8', src/main.rs:3:8
```

As some of the API between the compiler and the library landscape gets broken, this is a bit hackier than I'd originally wanted it to be.

* `panic` and `panic_bounds_check` lang items got an additional column param, on stage0 I still have to use the previous version. After a SNAP this should be resolved.
* For `#[derive(RustcDeserialze)]`, stage0 requires a fixed signature for `std::rt::begin_panic`, so we can't change it right away. What we need to do instead is to keep the signature, and add a `begin_panic_new` function that we use in later stages instead. After a SNAP we can change the `begin_panic` function and rely on it instead of `begin_panic_new`, and one SNAP later we can remove `begin_panic_new`.
* Fortunately I didn't have to break anything about the panic hook API, I could easily extend it.

Note that debuginfo remains unchanged, so RUST_BACKTRACE output won't contain any column info. See issue #42921 for discussion on including the column in debuginfo.

rust-highfive commented:
r? @brson

(rust_highfive has picked a reviewer for you, use r? to override)

eddyb commented:
cc @rust-lang/libs About the addition of `panicking::Location::column`. r=me otherwise.

est31 commented:
I've renamed the commit (without any code changes) and added a second one with a small refactor, a comment for the stage0 updater, and (hopefully) a fix for the test failure.

sfackler commented:
:+1:

est31 commented:
ready. r? @eddyb

est31 commented:
cc @alexcrichton

alexcrichton commented on src/libcore/panicking.rs:
```diff
@@ -39,34 +39,55 @@
         #[lang = "panic_fmt"]
         #[unwind]
-        fn panic_impl(fmt: fmt::Arguments, file: &'static str, line: u32) -> !;
+        fn panic_impl(fmt: fmt::Arguments, file: &'static str, line: u32, col :u32) -> !;
```
s/col :u32/col: u32/

est31 commented:
> Mind doing some size analysis to ensure this doesn't increase the size of anything too much?

I've compiled rustc with the patch and with the commit this PR bases on, and these are the results:

```

Without patch:
120136	build/x86_64-unknown-linux-gnu/stage1-rustc
93752	build/x86_64-unknown-linux-gnu/stage1-std

With patch:
120148	build/x86_64-unknown-linux-gnu/stage1-rustc
93772	build/x86_64-unknown-linux-gnu/stage1-std

```

Numbers come from du output, and are therefore in kb.

So almost no noticeable change, its in the sub-0.1% range for both std and rustc.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 376c6e7 has been approved by `eddyb`

<!-- @bors r=eddyb 376c6e71b0aa8bc62d5ff4a469d35a6feb063c0c -->

bors commented:
:hourglass: Testing commit 376c6e71b0aa8bc62d5ff4a469d35a6feb063c0c with merge 018afe2f0682464086a7ebae11637bf160a4a711...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/249273278?utm_source=github_status&utm_medium=notification)

est31 commented:
The failure looks legit, I'll investigate.

est31 commented:
The error is inside cargo's testsuite:

<details><summary>(expand)</summary>

```

[01:11:06] thread 'cargo_bench_failing_test' panicked at '
[01:11:06] Expected: execs
[01:11:06]     but: expected to find:
[01:11:06] [..]src[/]foo.rs:14
[01:11:06] 
[01:11:06] did not find in output:
[01:11:06]    Compiling foo v0.5.0 (file:///checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/cit/t11/foo)
[01:11:06]     Finished release [optimized] target(s) in 0.67 secs
[01:11:06]      Running target/release/deps/foo-b865dce48c3e4e89
[01:11:06] thread 'main' panicked at 'assertion failed: `(left == right)`
[01:11:06]   left: `"hello"`,
[01:11:06]  right: `"nope"`', src/foo.rs:14:16
[01:11:06] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:11:06] error: bench failed
[01:11:06] ', /cargo/registry/src/github.com-1ecc6299db9ec823/hamcrest-0.1.1/src/core.rs:31:12
```

</details>

est31 commented:
Fixed & rebased due to merge conflicts. re-r? @eddyb

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 4511a3a has been approved by `eddyb`

<!-- @bors r=eddyb 4511a3a64b1d805cfbc27397bb3b478227e8ddd3 -->

bors commented:
:hourglass: Testing commit 4511a3a64b1d805cfbc27397bb3b478227e8ddd3 with merge 1471dcdbe48193ebd47af265d6f8c38721462315...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/249311853?utm_source=github_status&utm_medium=notification)

Mark-Simulacrum commented:
I don't see an update to Cargo.toml in this PR, which I believe is needed -- we store the git sha for Cargo in there as well, IIRC.

```
[01:12:30] ---- cargo_test_failing_test_in_lib stdout ----
[01:12:30] 	running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/cargo test`
[01:12:30] thread 'cargo_test_failing_test_in_lib' panicked at '
[01:12:30] Expected: execs
[01:12:30]     but: expected to find:
[01:12:30] test test_hello ... FAILED
[01:12:30] 
[01:12:30] failures:
[01:12:30] 
[01:12:30] ---- test_hello stdout ----
[01:12:30] <tab>thread 'test_hello' panicked at 'assertion failed: false', src[/]lib.rs:4
[01:12:30] 
[01:12:30] 
[01:12:30] did not find in output:
[01:12:30] 
[01:12:30] running 1 test
[01:12:30] test test_hello ... FAILED
[01:12:30] 
[01:12:30] failures:
[01:12:30] 
[01:12:30] ---- test_hello stdout ----
[01:12:30] <tab>thread 'test_hello' panicked at 'assertion failed: false', src/lib.rs:4:16
[01:12:30] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:12:30] 
[01:12:30] 
[01:12:30] failures:
[01:12:30]     test_hello
[01:12:30] 
[01:12:30] test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
[01:12:30] 
[01:12:30] ', /cargo/registry/src/github.com-1ecc6299db9ec823/hamcrest-0.1.1/src/core.rs:31:12
[01:12:30] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:12:30] 
[01:12:30] ---- cargo_test_failing_test_in_test stdout ----
[01:12:30] 	running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/cargo build`
[01:12:30] running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/cit/t11/foo/target/debug/foo`
[01:12:30] running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/cargo test`
[01:12:30] thread 'cargo_test_failing_test_in_test' panicked at '
[01:12:30] Expected: execs
[01:12:30]     but: expected to find:
[01:12:30] running 1 test
[01:12:30] test test_hello ... FAILED
[01:12:30] 
[01:12:30] failures:
[01:12:30] 
[01:12:30] ---- test_hello stdout ----
[01:12:30] <tab>thread 'test_hello' panicked at 'assertion failed: false', tests[/]footest.rs:4
[01:12:30] 
[01:12:30] 
[01:12:30] did not find in output:
[01:12:30] 
[01:12:30] running 0 tests
[01:12:30] 
[01:12:30] test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
[01:12:30] 
[01:12:30] 
[01:12:30] running 1 test
[01:12:30] test test_hello ... FAILED
[01:12:30] 
[01:12:30] failures:
[01:12:30] 
[01:12:30] ---- test_hello stdout ----
[01:12:30] <tab>thread 'test_hello' panicked at 'assertion failed: false', tests/footest.rs:4:16
[01:12:30] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:12:30] 
[01:12:30] 
[01:12:30] failures:
[01:12:30]     test_hello
[01:12:30] 
[01:12:30] test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
[01:12:30] 
[01:12:30] ', /cargo/registry/src/github.com-1ecc6299db9ec823/hamcrest-0.1.1/src/core.rs:31:12
[01:12:30] 
[01:12:30] 
[01:12:30] failures:
[01:12:30]     cargo_test_failing_test_in_lib
[01:12:30]     cargo_test_failing_test_in_test
```

est31 commented:
@Mark-Simulacrum are you sure that an update to Cargo.toml/Cargo.lock is needed? E.g. look at git commit 77931c2a04fbb41f3cd92435aa5294222bb5db7f, it has no update.

About the other cargo failures... I'll try to fix them as well. Right now my hardware is a bit limited so I can't really run the testsuite sadly...

Mark-Simulacrum commented:
Cargo.lock was updated though: https://github.com/rust-lang/rust/commit/77931c2a04fbb41f3cd92435aa5294222bb5db7f#diff-5d9bf57935c52008319cccc88d4d737dR160.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 3b91f94 has been approved by `eddyb`

<!-- @bors r=eddyb 3b91f9406cd4b00ef09d19aca8203ce3539a202d -->

bors commented:
:hourglass: Testing commit 3b91f9406cd4b00ef09d19aca8203ce3539a202d with merge 0679711398bef656699e1ff6b004ecccbdb67284...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3781), [status-travis](https://travis-ci.org/rust-lang/rust/builds/249356597?utm_source=github_status&utm_medium=notification)
Approved by: eddyb
Pushing 0679711398bef656699e1ff6b004ecccbdb67284 to master...
